File: examples/goblins/behavior-wasi-c-context/worker.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static char *read_all(const char *path) {
    FILE *file = fopen(path, "rb");
    if (file == NULL) {
        fprintf(stderr, "could not open %s\n", path);
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    rewind(file);
    char *buffer = calloc((size_t)size + 1, sizeof(char));
    fread(buffer, 1, (size_t)size, file);
    fclose(file);
    return buffer;
}

static const char *env_or(const char *name, const char *fallback) {
    const char *value = getenv(name);
    return value == NULL || strlen(value) == 0 ? fallback : value;
}

int main(void) {
    char *input = read_all(getenv("GOBLIN_INPUT_PATH"));
    char *context = read_all(getenv("GOBLIN_CONTEXT_PATH"));
    printf("WASI C context goblin read %zu input bytes and %zu context bytes.\n", strlen(input), strlen(context));
    FILE *result = fopen(getenv("GOBLIN_RESULT_PATH"), "wb");
    fprintf(result,
            "{\n"
            "  \"status\": \"success\",\n"
            "  \"data\": {\n"
            "    \"message\": \"WASI context read complete\",\n"
            "    \"language\": \"c-wasi\",\n"
            "    \"run_id\": \"%s\",\n"
            "    \"input_bytes\": %zu,\n"
            "    \"context_bytes\": %zu\n"
            "  },\n"
            "  \"artifacts\": [],\n"
            "  \"metrics\": {\"input_bytes\": %zu, \"context_bytes\": %zu},\n"
            "  \"handoff\": [],\n"
            "  \"error\": null\n"
            "}\n",
            env_or("GOBLIN_RUN_ID", "unknown-run"),
            strlen(input),
            strlen(context),
            strlen(input),
            strlen(context));
    fclose(result);
    free(input);
    free(context);
    return 0;
}

File: examples/goblins/hello-python/worker.py
"""Minimal Python worker that implements the Goblin container contract."""

from __future__ import annotations

import json
import os
from pathlib import Path


def read_json_env(name: str) -> dict:
    path = Path(os.environ[name])
    with path.open("r", encoding="utf-8") as handle:
        value = json.load(handle)
    if not isinstance(value, dict):
        raise ValueError(f"{name} must point to a JSON object")
    return value


def main() -> None:
    input_data = read_json_env("GOBLIN_INPUT_PATH")
    read_json_env("GOBLIN_CONTEXT_PATH")
    result_path = Path(os.environ["GOBLIN_RESULT_PATH"])
    run_id = os.environ.get("GOBLIN_RUN_ID", "unknown-run")
    kind = os.environ.get("GOBLIN_KIND", "example.hello-python")
    target = input_data.get("target", "World")

    print(f"Python goblin says hello to {target}. The crown enjoys stdlib.")
    result = {
        "status": "success",
        "data": {
            "message": "Hello World",
            "language": "python",
            "runtime": "Python 3.12 standard library",
            "kind": kind,
            "run_id": run_id,
            "target": target,
            "input": input_data,
            "quote": "A goblin with no dependencies travels lightly.",
        },
        "artifacts": [],
        "metrics": {},
        "handoff": [],
        "error": None,
    }
    result_path.write_text(json.dumps(result, indent=2), encoding="utf-8")


if __name__ == "__main__":
    main()
